projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -374,13 +386,28 @@ private[projection] class R2dbcOffsetStore(
           newState
 
       R2dbcExecutor.updateInTx(stmts).map { _ =>
-        if (!state.compareAndSet(oldState, evictedNewState))
+        if (state.compareAndSet(oldState, evictedNewState))
+          cleanupSeen(evictedNewState)
+        else
           throw new IllegalStateException("Unexpected concurrent modification of state from saveOffset.")
         Done
       }
     }
   }
 
+  @tailrec private def cleanupSeen(newState: State): Unit = {
+    val currentSeen = getSeen()
+    val newSeen =
+      currentSeen.filter { case (seenPid, seenSeqNr) =>
+        newState.byPid.get(seenPid) match {
+          case Some(r) => r.seqNr < seenSeqNr
+          case None    => true
+        }
+      }
```

Comment:
One day I'll know filter vs filterNot by heart... So eh, keep all pid->seqnr in `seen` where there is no entry for the pid in the `newState` or where the new state has a lower sequence number than what was already in seen? *scratches head*

Reply:
that description matches the intended implementation